lint: Check for missing bitcoin-config.h includes

maflcko opened the pull request:
Missing `bitcoin-config.h` includes are problematic, because the build could silently pass, but produce an unintended result. For example, a slower fallback algorithm could be picked, even though `bitcoin-config.h` indicates that a faster feature is available and should be used.

As the build succeeds silently, this problem is not possible to detect with iwyu.

Thus, fix this by using a linter based on grepping the source code.

Commit:
lint: Add get_subtrees() helper

This is needed for a future change.

Commit:
lint: Check for missing or redundant bitcoin-config.h includes

theuni commented:
Concept ACK.

The linter output matches my list, with the additions of:
```
src/crypto/sha256_arm_shani.cpp
src/crypto/sha256_avx2.cpp
src/crypto/sha256_sse41.cpp
src/crypto/sha256_x86_shani.cpp
src/interfaces/wallet.h
```

The crypto defines actually come from our Makefile. Those have always been wonky, maybe now is a good time to do something about those.

`src/interfaces/wallet.h` however is a false-positive :(

TheCharlatan commented on test/lint/test_runner/src/main.rs:
```diff
@@ -87,6 +95,68 @@ fs:: namespace, which has unsafe filesystem functions marked as deleted.
+            .expect("command error")
+            .success());
+    }
+    let defines_regex = check_output(Command::new("grep").args(["undef ", "--", config_path]))
```
How about this to at least filter out the easy commented lines?
```diff
diff --git a/test/lint/test_runner/src/main.rs b/test/lint/test_runner/src/main.rs
index d9950c3fba..fa5ac0a597 100644
--- a/test/lint/test_runner/src/main.rs
+++ b/test/lint/test_runner/src/main.rs
@@ -106,10 +106,16 @@ fn lint_includes_build_config() -> LintResult {
-    let defines_regex = check_output(Command::new("grep").args(["undef ", "--", config_path]))
-        .expect("grep failed")
-        .lines()
-        .map(|line| {
-            line.split("undef ")
-                .nth(1)
-                .expect(&format!("Could not extract name in line: {line}"))
-        })
-        .collect::<Vec<_>>()
-        .join("|");
+    let defines_regex = format!(
+        "^(?!//).*({})",
+        check_output(Command::new("grep").args(["undef ", "--", config_path]))
+            .expect("grep failed")
+            .lines()
+            .map(|line| {
+                line.split("undef ")
+                    .nth(1)
+                    .expect(&format!("Could not extract name in line: {line}"))
+            })
+            .collect::<Vec<_>>()
+            .join("|")
+    );
+
+    println!("{}", defines_regex);
+
@@ -120 +126 @@ fn lint_includes_build_config() -> LintResult {
-                "--extended-regexp",
+                "--perl-regexp",
```

theuni replied:
I think this is a reasonable enough escape hatch. Given that we only have 1 currently, requiring a `// FOO` rather than a `/* FOO */` seems ok to me.

I don't see how we could do it as a clang-tidy plugin unfortunately. Its preprocessor stuff kinda sucks, and I believe we'd also have to keep a hard-coded list because any actual `bitcoin-config.h` would be guaranteed to leave _some_ things undef'd. Hmm, I suppose the plugin itself could read the `.h.in` for each invocation. That's pretty ugly though :(

epiccurious commented:
Concept ACK f4db3546c003822e485f1597701f278eef197619.

Commit:
doc: Add missing RUST_BACKTRACE=1

This will print a backtrace when an internal code error happened.

Commit:
lint: Make lint error easier to spot in output

theuni commented:
What's the status on this? Just waiting on a change for the crypto defines?

It'd be nice to get it in with/soonafter #29404.

hebasto reviewed (commented):
Tested 48562917172accd911e4c86f1032e4282fd321f5 on Ubuntu 23.10.

Missing `#include <config/bitcoin-config.h>` are reported correctly.

What are we supposed to do to automatically avoid redundant / unneeded includes?

hebasto commented:
> > What are we supposed to do to automatically avoid redundant / unneeded includes?
> 
> Iwyu is a tool to detect redundant and unneeded includes, so it can be used. However, there may be false warnings in regard to this specific header. Thus, an iwyu pragma can be used to avoid incorrect removal of needed includes. Should I push a commit here?

IWYU falsely suggests to remove `#include <config/bitcoin-config.h>`  in any case where the corresponding macro is commented out in the `config/bitcoin-config.h`.  For example, the `CHAR_EQUALS_INT8` macro in the `serialize.h`.

I suggest to add ` // IWYU pragma: keep` everywhere and put this rule to the [Developer Notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md).

hebasto commented:
> > What are we supposed to do to automatically avoid redundant / unneeded includes?
> 
> Iwyu is a tool to detect redundant and unneeded includes, so it can be used. However, there may be false warnings in regard to this specific header. Thus, an iwyu pragma can be used to avoid incorrect removal of needed includes. Should I push a commit here?

I assume that both macros `HAVE_TIMINGSAFE_BCMP` and `HAVE_EVHTTP_CONNECTION_GET_PEER_CONST_CHAR` are commented out in `config/bitcoin-config.h` in the CI environment.

I'm pretty sure that the same happens to the `CHAR_EQUALS_INT8` macro as well. However, IWYU does not suggest to drop `#include <config/bitcoin-config.h>` from the `serialize.h`. That seems fragile as IWYU behavior might change in the future..

I suggest to add  // IWYU pragma: keep everywhere and put this rule to the [Developer Notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md).

hebasto reviewed (commented):
https://api.cirrus-ci.com/v1/task/5330278671974400/logs/ci.log:

```
crypto/sha256.cpp should remove these lines:
- #include <config/bitcoin-config.h>  // lines 6-6
```

More `// IWYU pragma: keep`?

theuni commented:
It's not clear to me what the benefit of using IWYU is here. The missing defines will (by definition) vary per-platform, so I'm not sure what we're chasing here exactly.

Or is it more-so just to keep IWYU usable without warnings for our c-i configs specifically?

hebasto commented:
> Or is it more-so just to keep IWYU usable without warnings for our c-i configs specifically?

Yes. At some point in the future we might want to treat IWYU warnings as errors.

Commit:
refactor: Add missing include for USE_BDB, USE_SQLITE to bench/wallet_ismine.cpp

It was included indirectly via src/wallet/test/util.h, however it is
better to include what you use.

theuni reviewed (approved):
Weak ACK fa58ae74ea67485495dbc2d003adfbca68d6869b.

My python-fu is weak, but I think this is a necessary check and it seems to work as advertised.

sedited reviewed (approved):
ACK fa58ae74ea67485495dbc2d003adfbca68d6869b

hebasto reviewed (approved):
ACK fa58ae74ea67485495dbc2d003adfbca68d6869b, tested on Ubuntu 23.10 -- it catches bugs properly. I didn't review rust code changes.
